aws-s3-transfer-manager/src/operation/upload_objects/worker.rs

```diff
@@ -217,11 +255,19 @@ async fn upload_single_obj(
         .build()
         .expect("valid input");
 
-    let handle = crate::operation::upload::Upload::orchestrate(ctx.handle.clone(), input).await?;
-
-    handle.join().await?;
+    let mut handle =
+        crate::operation::upload::Upload::orchestrate(ctx.handle.clone(), input).await?;
 
-    Ok(bytes_transferred)
+    if cancel_rx.has_changed().unwrap() {
```

Comment:
Can we wrap this into a utility future like `MaybeCancel::new(fut)`? that checks cancelation on each poll? Feels like it would be useful.

Reply:
This particular `if` statement is technically optional (note `has_changed` is a sync method), and things would work just fine without it. It's just here so that we can potentially save a single upload operation that would otherwise be wasted if the world would already be in graceful shutdown mode. So a utility future may not be as useful as you might think (or I guess that depends on what utility future you were think of)?

Reply:
Fair enough, we can wait and abstract later as use cases arise (if they arise).

Reply:
Yes. For now, at least I added a comment in e655aeb based on my response above.